### PR TITLE
Refactor OutputsManager to align with default Jupyter behavior

### PR DESCRIPTION
**Pull request by ellisonbg**
## Summary

This PR refactors the notebook output handling system to align with standard Jupyter behavior while maintaining the performance benefits of the collaborative server architecture.

#### **OutputsManager Now Follows Default Jupyter Behavior**

The refactored `OutputsManager` now works exactly like standard Jupyter:
- **Full outputs saved to notebook files**: All cell outputs are written to `.ipynb` files on disk, ensuring compatibility with standard Jupyter workflows
- **Fast autosaving preserved**: Outputs are still kept out of the YDoc using lightweight placeholders, maintaining the performance advantage of server-side collaboration
- **No metadata flags required**: Works transparently without special notebook metadata settings

This means users get the best of both worlds:
- ✅ Standard Jupyter file format with full outputs on disk
- ✅ Fast collaborative editing with minimal YDoc synchronization overhead
- ✅ No breaking changes to existing workflows

#### **Dynamic Autosave Intervals**

The `YRoomFileAPI` now implements adaptive autosave timing to optimize performance across different file sizes and I/O environments:

- **Adaptive timing**: Poll interval automatically adjusts based on how long saves take
  - Small files with fast I/O → shorter intervals → more responsive autosaving
  - Large files or slower I/O → longer intervals → reduces unnecessary polling overhead
- **Configurable parameters**:
  - `min_poll_interval` (default: 0.5s) - minimum autosave interval
  - `poll_interval_multiplier` (default: 5.0) - multiplier applied to save duration to calculate next interval
- **Example**: If a save takes 2 seconds, the next poll interval becomes `2s × 5 = 10s`, scaling appropriately to the file size and I/O performance

This ensures the autosave system adapts to realistic file I/O conditions, providing optimal responsiveness for small files while avoiding excessive polling overhead for large files or slower storage systems.

#### **Cleaner, Better-Documented Code**

- **Comprehensive documentation**: All methods now have detailed docstrings explaining purpose, parameters, and return values
- **Better code organization**: Extracted utility functions (`_create_output_url`, `_create_output_placeholder`) and helper methods (`_upgrade_notebook_format`, `_ensure_cell_id`, `_process_outputs_from_cell`)
- **Simplified logic**: Removed complex conditional paths, making the code easier to understand and maintain
- **Improved error handling**: Graceful fallbacks when outputs aren't found

#### **Expanded Test Coverage**

- Comprehensive test suite (857 lines) covering all output types
- Tests for display ID tracking, notebook loading/saving workflows
- Edge case and error handling validation

### Experimental Features

⚠️ **Note**: This PR also includes an experimental `OptimizedOutputsManager` that supports excluding outputs from saved notebook files via an `exclude_outputs` metadata flag. This feature is **disabled by default** and not recommended for production use.

### Migration Impact

**No migration needed** - existing notebooks and workflows continue to work unchanged. The refactored OutputsManager is a drop-in replacement with improved code quality and documentation.

### CODE REVIEW

File: jupyter_server_documents/rooms/yroom_file_api.py [hunk 1]
```diff
@@ -95,7 +126,28 @@ class YRoomFileAPI(LoggingConfigurable):
     dual-writes or dirty-reads.
     """
 
+    _last_save_duration: float | None
+    """
+    The duration in seconds of the last save operation. Used to calculate the
+    adaptive poll interval.
+    """
```
Comment (dlqqq) [inline jupyter_server_documents/rooms/yroom_file_api.py]: This instance attribute doesn't seem necessary, since it is only bound to the local variable `save_duration` in the `save()` method.

**Comment by 3coins** [comment]
Here is a comparison of current algorithm with exponential moving average for some hypothetical notebook save times.

5d3aca12-3949-45e0-a845-8e363e7a5e62

**Comment by dlqqq** [comment]
@Zsailer Thanks for letting us know. I've opened https://github.com/jupyter-ai-contrib/jupyter-server-documents/pull/168, which implements the feedback that I had given earlier. Once that is merged into this branch, we can merge this PR to unblock you.

I can also publish a release of this afterwards to allow us to verify these changes.

**Comment by dlqqq** [comment]
Oh, that PR is failing CI badly because this branch needs a rebase first. I can help with that.

Actually, it is easier to merge this as-is & rebase #168 onto `main` after this is merged.

**Comment by dlqqq** [comment]
@Zsailer This PR is now merged, so you're clear to open your PR! 🎉
